torch/nn/functional.py

```diff
@@ -2259,7 +2259,7 @@ def soft_margin_loss(input, target, size_average=None, reduce=None, reduction='m
         reduction_enum = _Reduction.legacy_get_enum(size_average, reduce)
     else:
         reduction_enum = _Reduction.get_enum(reduction)
-    return torch._C._nn.soft_margin_loss(input, target, reduction_enum)
+    return torch.soft_margin_loss(input, target, reduction_enum)
```

Comment:
I was not sure what the general rules are for the c-function imports in python. I followed the structure of the other functions in [Loss.cpp](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/Loss.cpp). Particularly the entry in native_functions.yaml now looks like:
```
- func: soft_margin_loss(Tensor self, Tensor target, int reduction=Mean) -> Tensor
  use_c10_dispatcher: full
```
e.g. the function becomes accessible in Python as `torch.soft_margin_loss()` .. is this fine or what are the rules/recommendations here?

Reply:
Umm... it ought not to but I took a look at the code and there's no way to do this right now. I think ideally you could declare this as `func: nn::soft_margin_loss...` and then it would go in torch.nn namespace. So this is OK for now.

Reply:
It would be possible to add the `python_module: nn` line to native_functions.yaml. Then the function would become visible in python as `torch._C._nn.soft_margin_loss` (as it was before). But unless there is a special reason to have the other loss functions somewhere else it should be the same for all functions that are currently defined in Loss.cpp. As far as I understood for 'normal' nn modules the `python_module: nn` should be used, e.g. as done in other ATen PRs.
Nobody should be directly using the 'C' functions imported directly to `torch.xyz' ... so I think this could be changed easily.

@t-vi could you please explain what the following comment in gen_pyi.py means?
https://github.com/pytorch/pytorch/blob/341262754f5e7247e136504fcd06bf6fcab0dcc9/tools/pyi/gen_pyi.py#L59 Do you know why the loss functions are not residing under `torch._C.nn.` ?

Reply:
I think the git history is a bit coarse here. The comment probably was from Ed.
But what it means is that these functions are wrapped and the reduction enum from C++ is not exposed to Python (and so we don't know how to/don't want to translate it in the pyi generation).

In general, I think it would be preferrable to not change the user-facing signature (which I would count the nn attribute to which the patch removes).

Reply:
@t-vi ok, we can leave the function imported to _C.nn. What do you think about the other loss functions? Shouldn't they also be moved to _C.nn? And should we keep the variants with out-tensor argument? If yes, should they be added for the ones which currently do not have them?

Reply:
So I have not looked closely at the TH to ATen transitions, but I would think that we keep `.out` variants if we have them right now and keep the functions sorted in `_C._nn`, too. (Whether one might prune the public API is an interesting question, but as you I think that here we are not looking for BC breaking changes in here, and removing things is obviously BC breaking. My sympathy here does not extend to people doing funny things with `_backward`, so it would seem to be OK to remove those.)
I don't know whether `.out` functions for all losses are a priority, we did add them for most "basic" functions but there was also some need for them.
I must admit that I am ignorant of whether moving functions exposed via torch.nn.functional functions to `_C._nn` is a good idea: it seems a reasonable cleanup to me, but I am not quite sure whether that would wreck havoc for the non-Python bindings, so probably you'd need someone else to weight in on that. (But it's not for this patch.)

Reply:
@t-vi @ezyang I implemented inplace forward & backward functions for `soft_margin_loss`. I am able to call the forward version of it using `torch._C._nn.soft_margin_loss(a, b, reduction=0, out=c)` ... but I am not able to trigger *any* of the TH out-style loss backward function, I tried for example to use mse_loss as a reference:

```
a = torch.rand(10, requires_grad=True)
b = torch.ones(10)
c = torch.empty(10)
torch._C._nn.mse_loss(a, b, reduction=1, out=c)
```
which triggers th error `RuntimeError: mse_loss(): functions with out=... arguments don't support automatic differentiation, but one of the arguments requires grad.`

I also tried with other loss-functions, e.g. `torch._C._nn.l1_loss(a, b, reduction=1, out=c)`, `torch._C._nn.smooth_l1_loss(a, b, reduction=1, out=c)` ... all trigger the same error.

I do not find any special entries in `derivatives.yaml` for the old TH out-variants. Are they really still used and accessible from python?

Reply:
That's a general limitation not specific to loss functions or old-style TH functions. Any function with `out` kwarg cannot be automatically differentiated.

Reply:
@ngimel Sorry, I feel I miss here some important bits. Probably a good first question to clarify things for me would be: How can I invoke the inplace-backward variants (e.g. the `xyz.grad_input()` entries) in [native_functions.yaml](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/native_functions.yaml#L5699) from python?

Reply:
Probably there's no way and your initial hypothesis that they are not accessible from python is correct? I'm not an authority on this, though, I just know I've been bitten by inability to compute gradients for functions with `out` arg in the past. Also, in-place backward functions (i.e. those where gradInput overwrites gradOutput) are in general frowned upon, making it very likely that they are not really exposed anywhere. Rationale here is if a few variables share the same gradient (situation arising e.g. when there is `c=a+b` in the forward path, `a` and `b` gradients would reference the same data), and then one of them in-place modifies it, this situation is hard to handle properly.